server/src/main/java/org/opensearch/rest/pagination/IndexBasedPaginationStrategy.java

```diff
@@ -0,0 +1,191 @@
+            // find the index whose creation time is just after/before (based on sortOrder) the index which was last responded.
+            while (requestedPageStartIndexNumber > 0) {
+                long creationTimeOfIndex = clusterState.metadata()
+                    .indices()
+                    .get(sortedIndicesList.get(requestedPageStartIndexNumber - 1))
+                    .getCreationDate();
+                if ((!DESCENDING_SORT_PARAM_VALUE.equals(sortOrder)
+                    && creationTimeOfIndex < requestedTokenParser.creationTimeOfLastRespondedIndex)
+                    || (DESCENDING_SORT_PARAM_VALUE.equals(sortOrder)
+                        && creationTimeOfIndex > requestedTokenParser.creationTimeOfLastRespondedIndex)) {
+                    break;
+                }
+                requestedPageStartIndexNumber--;
+            }
+        }
+        return requestedPageStartIndexNumber;
+    }
+
+    /**
+     * TokenParser to be used by {@link IndexBasedPaginationStrategy}.
+     * Token would like: IndexNumberToStartTheNextPageFrom + $ + CreationTimeOfLastRespondedIndex + $ +
+     * QueryStartTime + $ + NameOfLastRespondedIndex
+     */
+    public static class IndexStrategyTokenParser {
+
+        private final int posToStartPage;
+        private final long creationTimeOfLastRespondedIndex;
+        private final long queryStartTime;
+        private final String nameOfLastRespondedIndex;
+
+        public IndexStrategyTokenParser(String requestedTokenString) {
+            validateIndexStrategyToken(requestedTokenString);
+            String decryptedToken = PaginationStrategy.decryptStringToken(requestedTokenString);
+            final String[] decryptedTokenElements = decryptedToken.split("\\$");
+            this.posToStartPage = Integer.parseInt(decryptedTokenElements[0]);
+            this.creationTimeOfLastRespondedIndex = Long.parseLong(decryptedTokenElements[1]);
+            this.queryStartTime = Long.parseLong(decryptedTokenElements[2]);
+            this.nameOfLastRespondedIndex = decryptedTokenElements[3];
+        }
+
+        public IndexStrategyTokenParser(
+            int indexNumberToStartPageFrom,
+            long creationTimeOfLastRespondedIndex,
+            long queryStartTime,
+            String nameOfLastRespondedIndex
+        ) {
+            Objects.requireNonNull(nameOfLastRespondedIndex, "index name should be provided");
+            this.posToStartPage = indexNumberToStartPageFrom;
+            this.creationTimeOfLastRespondedIndex = creationTimeOfLastRespondedIndex;
+            this.queryStartTime = queryStartTime;
+            this.nameOfLastRespondedIndex = nameOfLastRespondedIndex;
+        }
+
+        public String generateEncryptedToken() {
+            return PaginationStrategy.encryptStringToken(
+                posToStartPage + "$" + creationTimeOfLastRespondedIndex + "$" + queryStartTime + "$" + nameOfLastRespondedIndex
```

Comment:
- could the `nameOfLastRespondedIndex` contain the seperator `$` ?
- can we define constants to map the position of token to fields ?

Reply:
> could the nameOfLastRespondedIndex contain the seperator $ ?

Yeah, sorry this was one of the open point. "$" is an allowed index name, and was kept as a placeholder.